schemas/advanced-layouts.md

```diff
@@ -0,0 +1,120 @@
+Advanced Layouts for IPLD Schemas
+---------------------------------
+
+An "advanced layout" is a special type in an IPLD Schema that denotes an abstract node that may _act_ as one of the standard IPLD [schema kinds](./schema-kinds.md) when interacted with above the schema layer, but be represented in entirely different data forms at the data model layer. Advanced layouts allow us to use transformational logic to present consistent interfaces to significantly more complex data than the data model otherwise allows or significantly larger data than reasonable block sizes allow.
+
+Data structures such as maps, lists or byte arrays may be implemented in novel forms with the data model and may even transparently cross block boundaries to present data that is potentially both larger and more complex than a standard data model kind while affording the standard data model API interaction modes of their respective kinds.
+
+Example uses of advanced layouts include:
+
+* Very large byte arrays that may cross many block boundaries but still present a standard `bytes` kind interface.
+* Maps that can be sharded across many blocks allowing efficient access via standard `map` kind interfaces while allowing extremely large storage.
+* Encrypted block mechanisms that are able to encode and decode data at the data model layer (perhaps in `byte` or `string` form) while presenting as unencrypted kinds when accessed through the schema layer.
+* Novel kind representation formats, such as high-precision or large numbers or alternative string encoding formats. Although size constraints at the language layer for number kinds may represent challenges for large or high-precision number formats.
+
+Advanced layouts necessarily involve a form of logic to perform their data transformations. Many use-cases even necessitate interaction with block loading and storing mechanisms, such as traversal of a large `bytes` array stored across a chain of blocks. There is currently no formal requirement for where this logic resides, how it is loaded, or how it is associated with an IPLD Schema definition. This topic will continue to evolve. There is potential in the future for such logic to be embedded _in_ IPLD blocks themselves via WebAssembly, such that an advanced layout is represented both by the data and the logic required to read it. See [#130](https://github.com/ipld/specs/issues/130) for an early exploration of some of these themes.
+
+## Basic schema definition and use
+
+At its most basic, no properties are mandated for advanced layout. One may be defined simply with the keyword `advanced` in a similar manner to `type`:
+
+```ipldsch
+advanced ROT13
+```
+
+It may then be used as a `representation` for a `type`, from which we can infer the `kind` that we expect.
+
+```ipldsch
+type MyString string representation ROT13
```

Comment:
Okay, so, I'm sold on using some specialization of the representation clause on the grounds that something certainly can't have another meaningful representation directive if it also has an advanced layout.

Two other issues with that, though:

- Needs a separate name plane.  (Thinking about how this should encode in the schema-schema *as json* rather than DSL may be helpful here.)  Two reasons: 1, it terrifies me that advanced layouts would be visually indistinctive on the same line they're used (I value lexical-locality highly for ease of fast reading), so let's avoid that by adding a sigil or something; 2, we don't want built-in representations and user-named advanced layouts to be able to get into name conflicts.  Implication: the current schema-schema json is also probably Insufficient to describe advanced layouts -- one possible solution is to turn the "representation" field there into *another* union of "core"|"advanced", and move the current unions to be under "core".

- Because this uses the `type` clause, this only works for named types.  Doesn't answer what we'd do if there's an inline usage.  E.g. `type Foo struct { wizardField {X:Y}<ShardedMap> }`.  Maaaaybe this is actually a usage we can just... not.  But it's worth a thought.  (For a fuller example using this syntax: [this unixfs schema spike](https://github.com/ipld/specs/blob/01f19b1f4234a9528c618756b7e9c24b4d499b7c/design/history/exploration-reports/2019.06-unixfsv2-spike-01.md#schema) uses inline markers for `Dir.members` and `File.content` -- it's interesting to note that this is 100% of the usages; and in both cases, the wrapper types would've existed and needed names anyway (for their relationship to unions), so requiring a named type for the advanced layout to be applied would have strictly increased the number of types needed.)

Reply:
Hrm, I'll take my own wind back out of my sails a bit for that second point.  Does having a named type means we can do better error messages?  Maybe.  Will codegen in practice  often need to produce a type for this node and thus need to name it something?  Uff, probably.  Can we solve those by munging a field name?  I guess, but it's not necessarily good looking.  Needs thought.

Reply:
... turning the schema-schema json into a a union nested in a union sounds gross, so maybe a better idea there would be to add a member to all the current representation unions that's just called "goto_adv" and has a string saying the name of the advanced layout; and then the top level schema objects should now contain a map of types and also a map of advancedlayout details.  That'd be a lot less twisty than a union-in-a-union, and maps well to the DSL syntax, and also still demonstrates the point that advanced layouts are Distinct in name plane from other representations.